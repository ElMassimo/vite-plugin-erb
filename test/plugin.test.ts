import { readFileSync } from 'fs'
import { join } from 'path'
import execa from 'execa'
import glob from 'fast-glob'
import type { InlineConfig } from 'vite'
import { describe, test, expect } from 'vitest'

type Fixture = 'a' | 'b' | 'c'

function compiledFile (name: Fixture) {
  const distPath = join(__dirname, 'fixtures', 'blog', 'public', name)
  const files = glob.sync('assets/application.*.js', { cwd: distPath })
  return readFileSync(join(distPath, files[0]), { encoding: 'utf8' })
}

async function buildFixture (name: Fixture, { mode }: InlineConfig) {
  await execa('bin/vite', ['build', '--clear', '--mode', mode], {
    cwd: join(__dirname, 'fixtures', 'blog'),
    env: {
      VITE_RUBY_ENTRYPOINTS_DIR: `entrypoints_${name}`,
      VITE_RUBY_PUBLIC_OUTPUT_DIR: name,
    },
  })
}

describe('erb', () => {
  test('javascript and css', async () => {
    await buildFixture('a', { mode: 'production' })
    const result = compiledFile('a')
    expect(result).toContain('"production"')
    expect(result).toMatchSnapshot()
  }, 10000)

  test('typescript and sass', async () => {
    await buildFixture('b', { mode: 'development' })
    const result = compiledFile('b')
    expect(result).toContain('"development"')
    expect(result).toMatchSnapshot()
  }, 10000)

  test('jsx', async () => {
    await buildFixture('c', { mode: 'test' })
    const result = compiledFile('c')
    expect(result).toContain('"test"')
    expect(result).toMatchSnapshot()
  }, 10000)
})
