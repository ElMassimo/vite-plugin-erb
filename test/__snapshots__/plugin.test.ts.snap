// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`erb > javascript and css 1`] = `
"const o="production",n="production";console.log({rackEnv:o,railsEnv:n});
//# sourceMappingURL=application-7Hm-zZ8S.js.map
"
`;

exports[`erb > jsx 1`] = `
"var fe={exports:{}},le={};/*
object-assign
(c) Sindre Sorhus
@license MIT
*/var ve,ar;function st(){if(ar)return ve;ar=1;var s=Object.getOwnPropertySymbols,_=Object.prototype.hasOwnProperty,G=Object.prototype.propertyIsEnumerable;function S(b){if(b==null)throw new TypeError("Object.assign cannot be called with null or undefined");return Object(b)}function $(){try{if(!Object.assign)return!1;var b=new String("abc");if(b[5]="de",Object.getOwnPropertyNames(b)[0]==="5")return!1;for(var C={},p=0;p<10;p++)C["_"+String.fromCharCode(p)]=p;var w=Object.getOwnPropertyNames(C).map(function(g){return C[g]});if(w.join("")!=="0123456789")return!1;var m={};return"abcdefghijklmnopqrst".split("").forEach(function(g){m[g]=g}),Object.keys(Object.assign({},m)).join("")==="abcdefghijklmnopqrst"}catch{return!1}}return ve=$()?Object.assign:function(b,C){for(var p,w=S(b),m,g=1;g<arguments.length;g++){p=Object(arguments[g]);for(var k in p)_.call(p,k)&&(w[k]=p[k]);if(s){m=s(p);for(var T=0;T<m.length;T++)G.call(p,m[T])&&(w[m[T]]=p[m[T]])}}return w},ve}var or;function ct(){return or||(or=1,function(s){/** @license React v17.0.2
 * react.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */(function(){var _=st(),G="17.0.2",S=60103,$=60106;s.Fragment=60107,s.StrictMode=60108,s.Profiler=60114;var b=60109,C=60110,p=60112;s.Suspense=60113;var w=60120,m=60115,g=60116,k=60121,T=60122,de=60117,pe=60129,he=60131;if(typeof Symbol=="function"&&Symbol.for){var d=Symbol.for;S=d("react.element"),$=d("react.portal"),s.Fragment=d("react.fragment"),s.StrictMode=d("react.strict_mode"),s.Profiler=d("react.profiler"),b=d("react.provider"),C=d("react.context"),p=d("react.forward_ref"),s.Suspense=d("react.suspense"),w=d("react.suspense_list"),m=d("react.memo"),g=d("react.lazy"),k=d("react.block"),T=d("react.server.block"),de=d("react.fundamental"),d("react.scope"),d("react.opaque.id"),pe=d("react.debug_trace_mode"),d("react.offscreen"),he=d("react.legacy_hidden")}var me=typeof Symbol=="function"&&Symbol.iterator,ir="@@iterator";function ge(e){if(e===null||typeof e!="object")return null;var r=me&&e[me]||e[ir];return typeof r=="function"?r:null}var ye={current:null},sr={transition:0},O={current:null},N={},Y=null;function be(e){Y=e}N.setExtraStackFrame=function(e){Y=e},N.getCurrentStack=null,N.getStackAddendum=function(){var e="";Y&&(e+=Y);var r=N.getCurrentStack;return r&&(e+=r()||""),e};var cr={current:!1},U={ReactCurrentDispatcher:ye,ReactCurrentBatchConfig:sr,ReactCurrentOwner:O,IsSomeRendererActing:cr,assign:_};U.ReactDebugCurrentFrame=N;function W(e){{for(var r=arguments.length,t=new Array(r>1?r-1:0),n=1;n<r;n++)t[n-1]=arguments[n];Re("warn",e,t)}}function c(e){{for(var r=arguments.length,t=new Array(r>1?r-1:0),n=1;n<r;n++)t[n-1]=arguments[n];Re("error",e,t)}}function Re(e,r,t){{var n=U.ReactDebugCurrentFrame,a=n.getStackAddendum();a!==""&&(r+="%s",t=t.concat([a]));var i=t.map(function(u){return""+u});i.unshift("Warning: "+r),Function.prototype.apply.call(console[e],console,i)}}var _e={};function B(e,r){{var t=e.constructor,n=t&&(t.displayName||t.name)||"ReactClass",a=n+"."+r;if(_e[a])return;c("Can't call %s on a component that is not yet mounted. This is a no-op, but it might indicate a bug in your application. Instead, assign to \`this.state\` directly or define a \`state = {};\` class property with the desired state in the %s component.",r,n),_e[a]=!0}}var Ee={isMounted:function(e){return!1},enqueueForceUpdate:function(e,r,t){B(e,"forceUpdate")},enqueueReplaceState:function(e,r,t,n){B(e,"replaceState")},enqueueSetState:function(e,r,t,n){B(e,"setState")}},X={};Object.freeze(X);function A(e,r,t){this.props=e,this.context=r,this.refs=X,this.updater=t||Ee}A.prototype.isReactComponent={},A.prototype.setState=function(e,r){if(!(typeof e=="object"||typeof e=="function"||e==null))throw Error("setState(...): takes an object of state variables to update or a function which returns an object of state variables.");this.updater.enqueueSetState(this,e,r,"setState")},A.prototype.forceUpdate=function(e){this.updater.enqueueForceUpdate(this,e,"forceUpdate")};{var J={isMounted:["isMounted","Instead, make sure to clean up subscriptions and pending requests in componentWillUnmount to prevent memory leaks."],replaceState:["replaceState","Refactor your code to use setState instead (see https://github.com/facebook/react/issues/3236)."]},fr=function(e,r){Object.defineProperty(A.prototype,e,{get:function(){W("%s(...) is deprecated in plain JavaScript React classes. %s",r[0],r[1])}})};for(var Q in J)J.hasOwnProperty(Q)&&fr(Q,J[Q])}function Ce(){}Ce.prototype=A.prototype;function Z(e,r,t){this.props=e,this.context=r,this.refs=X,this.updater=t||Ee}var ee=Z.prototype=new Ce;ee.constructor=Z,_(ee,A.prototype),ee.isPureReactComponent=!0;function lr(){var e={current:null};return Object.seal(e),e}function vr(e,r,t){var n=r.displayName||r.name||"";return e.displayName||(n!==""?t+"("+n+")":t)}function we(e){return e.displayName||"Context"}function j(e){if(e==null)return null;if(typeof e.tag=="number"&&c("Received an unexpected object in getComponentName(). This is likely a bug in React. Please file an issue."),typeof e=="function")return e.displayName||e.name||null;if(typeof e=="string")return e;switch(e){case s.Fragment:return"Fragment";case $:return"Portal";case s.Profiler:return"Profiler";case s.StrictMode:return"StrictMode";case s.Suspense:return"Suspense";case w:return"SuspenseList"}if(typeof e=="object")switch(e.$$typeof){case C:var r=e;return we(r)+".Consumer";case b:var t=e;return we(t._context)+".Provider";case p:return vr(e,e.render,"ForwardRef");case m:return j(e.type);case k:return j(e._render);case g:{var n=e,a=n._payload,i=n._init;try{return j(i(a))}catch{return null}}}return null}var V=Object.prototype.hasOwnProperty,Pe={key:!0,ref:!0,__self:!0,__source:!0},Oe,je,re;re={};function ke(e){if(V.call(e,"ref")){var r=Object.getOwnPropertyDescriptor(e,"ref").get;if(r&&r.isReactWarning)return!1}return e.ref!==void 0}function Te(e){if(V.call(e,"key")){var r=Object.getOwnPropertyDescriptor(e,"key").get;if(r&&r.isReactWarning)return!1}return e.key!==void 0}function dr(e,r){var t=function(){Oe||(Oe=!0,c("%s: \`key\` is not a prop. Trying to access it will result in \`undefined\` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)",r))};t.isReactWarning=!0,Object.defineProperty(e,"key",{get:t,configurable:!0})}function pr(e,r){var t=function(){je||(je=!0,c("%s: \`ref\` is not a prop. Trying to access it will result in \`undefined\` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)",r))};t.isReactWarning=!0,Object.defineProperty(e,"ref",{get:t,configurable:!0})}function hr(e){if(typeof e.ref=="string"&&O.current&&e.__self&&O.current.stateNode!==e.__self){var r=j(O.current.type);re[r]||(c('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref',r,e.ref),re[r]=!0)}}var te=function(e,r,t,n,a,i,u){var o={$$typeof:S,type:e,key:r,ref:t,props:u,_owner:i};return o._store={},Object.defineProperty(o._store,"validated",{configurable:!1,enumerable:!1,writable:!0,value:!1}),Object.defineProperty(o,"_self",{configurable:!1,enumerable:!1,writable:!1,value:n}),Object.defineProperty(o,"_source",{configurable:!1,enumerable:!1,writable:!1,value:a}),Object.freeze&&(Object.freeze(o.props),Object.freeze(o)),o};function mr(e,r,t){var n,a={},i=null,u=null,o=null,f=null;if(r!=null){ke(r)&&(u=r.ref,hr(r)),Te(r)&&(i=""+r.key),o=r.__self===void 0?null:r.__self,f=r.__source===void 0?null:r.__source;for(n in r)V.call(r,n)&&!Pe.hasOwnProperty(n)&&(a[n]=r[n])}var l=arguments.length-2;if(l===1)a.children=t;else if(l>1){for(var v=Array(l),h=0;h<l;h++)v[h]=arguments[h+2];Object.freeze&&Object.freeze(v),a.children=v}if(e&&e.defaultProps){var R=e.defaultProps;for(n in R)a[n]===void 0&&(a[n]=R[n])}if(i||u){var y=typeof e=="function"?e.displayName||e.name||"Unknown":e;i&&dr(a,y),u&&pr(a,y)}return te(e,i,u,o,f,O.current,a)}function gr(e,r){var t=te(e.type,r,e.ref,e._self,e._source,e._owner,e.props);return t}function yr(e,r,t){if(e==null)throw Error("React.cloneElement(...): The argument must be a React element, but you passed "+e+".");var n,a=_({},e.props),i=e.key,u=e.ref,o=e._self,f=e._source,l=e._owner;if(r!=null){ke(r)&&(u=r.ref,l=O.current),Te(r)&&(i=""+r.key);var v;e.type&&e.type.defaultProps&&(v=e.type.defaultProps);for(n in r)V.call(r,n)&&!Pe.hasOwnProperty(n)&&(r[n]===void 0&&v!==void 0?a[n]=v[n]:a[n]=r[n])}var h=arguments.length-2;if(h===1)a.children=t;else if(h>1){for(var R=Array(h),y=0;y<h;y++)R[y]=arguments[y+2];a.children=R}return te(e.type,i,u,o,f,l,a)}function D(e){return typeof e=="object"&&e!==null&&e.$$typeof===S}var Se=".",br=":";function Rr(e){var r=/[=:]/g,t={"=":"=0",":":"=2"},n=e.replace(r,function(a){return t[a]});return"$"+n}var Ae=!1,_r=/\\/+/g;function De(e){return e.replace(_r,"$&/")}function ne(e,r){return typeof e=="object"&&e!==null&&e.key!=null?Rr(""+e.key):r.toString(36)}function L(e,r,t,n,a){var i=typeof e;(i==="undefined"||i==="boolean")&&(e=null);var u=!1;if(e===null)u=!0;else switch(i){case"string":case"number":u=!0;break;case"object":switch(e.$$typeof){case S:case $:u=!0}}if(u){var o=e,f=a(o),l=n===""?Se+ne(o,0):n;if(Array.isArray(f)){var v="";l!=null&&(v=De(l)+"/"),L(f,r,v,"",function(it){return it})}else f!=null&&(D(f)&&(f=gr(f,t+(f.key&&(!o||o.key!==f.key)?De(""+f.key)+"/":"")+l)),r.push(f));return 1}var h,R,y=0,E=n===""?Se:n+br;if(Array.isArray(e))for(var K=0;K<e.length;K++)h=e[K],R=E+ne(h,K),y+=L(h,r,t,R,a);else{var ce=ge(e);if(typeof ce=="function"){var rr=e;ce===rr.entries&&(Ae||W("Using Maps as children is not supported. Use an array of keyed ReactElements instead."),Ae=!0);for(var ot=ce.call(rr),tr,ut=0;!(tr=ot.next()).done;)h=tr.value,R=E+ne(h,ut++),y+=L(h,r,t,R,a)}else if(i==="object"){var nr=""+e;throw Error("Objects are not valid as a React child (found: "+(nr==="[object Object]"?"object with keys {"+Object.keys(e).join(", ")+"}":nr)+"). If you meant to render a collection of children, use an array instead.")}}return y}function x(e,r,t){if(e==null)return e;var n=[],a=0;return L(e,n,"","",function(i){return r.call(t,i,a++)}),n}function Er(e){var r=0;return x(e,function(){r++}),r}function Cr(e,r,t){x(e,function(){r.apply(this,arguments)},t)}function wr(e){return x(e,function(r){return r})||[]}function Pr(e){if(!D(e))throw Error("React.Children.only expected to receive a single React element child.");return e}function Or(e,r){r===void 0?r=null:r!==null&&typeof r!="function"&&c("createContext: Expected the optional second argument to be a function. Instead received: %s",r);var t={$$typeof:C,_calculateChangedBits:r,_currentValue:e,_currentValue2:e,_threadCount:0,Provider:null,Consumer:null};t.Provider={$$typeof:b,_context:t};var n=!1,a=!1,i=!1;{var u={$$typeof:C,_context:t,_calculateChangedBits:t._calculateChangedBits};Object.defineProperties(u,{Provider:{get:function(){return a||(a=!0,c("Rendering <Context.Consumer.Provider> is not supported and will be removed in a future major release. Did you mean to render <Context.Provider> instead?")),t.Provider},set:function(o){t.Provider=o}},_currentValue:{get:function(){return t._currentValue},set:function(o){t._currentValue=o}},_currentValue2:{get:function(){return t._currentValue2},set:function(o){t._currentValue2=o}},_threadCount:{get:function(){return t._threadCount},set:function(o){t._threadCount=o}},Consumer:{get:function(){return n||(n=!0,c("Rendering <Context.Consumer.Consumer> is not supported and will be removed in a future major release. Did you mean to render <Context.Consumer> instead?")),t.Consumer}},displayName:{get:function(){return t.displayName},set:function(o){i||(W("Setting \`displayName\` on Context.Consumer has no effect. You should set it directly on the context with Context.displayName = '%s'.",o),i=!0)}}}),t.Consumer=u}return t._currentRenderer=null,t._currentRenderer2=null,t}var jr=-1,ae=0,Ie=1,kr=2;function Tr(e){if(e._status===jr){var r=e._result,t=r(),n=e;n._status=ae,n._result=t,t.then(function(a){if(e._status===ae){var i=a.default;i===void 0&&c(\`lazy: Expected the result of a dynamic import() call. Instead received: %s

Your code should look like: 
  const MyComponent = lazy(() => import('./MyComponent'))\`,a);var u=e;u._status=Ie,u._result=i}},function(a){if(e._status===ae){var i=e;i._status=kr,i._result=a}})}if(e._status===Ie)return e._result;throw e._result}function Sr(e){var r={_status:-1,_result:e},t={$$typeof:g,_payload:r,_init:Tr};{var n,a;Object.defineProperties(t,{defaultProps:{configurable:!0,get:function(){return n},set:function(i){c("React.lazy(...): It is not supported to assign \`defaultProps\` to a lazy component import. Either specify them where the component is defined, or create a wrapping component around it."),n=i,Object.defineProperty(t,"defaultProps",{enumerable:!0})}},propTypes:{configurable:!0,get:function(){return a},set:function(i){c("React.lazy(...): It is not supported to assign \`propTypes\` to a lazy component import. Either specify them where the component is defined, or create a wrapping component around it."),a=i,Object.defineProperty(t,"propTypes",{enumerable:!0})}}})}return t}function Ar(e){e!=null&&e.$$typeof===m?c("forwardRef requires a render function but received a \`memo\` component. Instead of forwardRef(memo(...)), use memo(forwardRef(...))."):typeof e!="function"?c("forwardRef requires a render function but was given %s.",e===null?"null":typeof e):e.length!==0&&e.length!==2&&c("forwardRef render functions accept exactly two parameters: props and ref. %s",e.length===1?"Did you forget to use the ref parameter?":"Any additional parameter will be undefined."),e!=null&&(e.defaultProps!=null||e.propTypes!=null)&&c("forwardRef render functions do not support propTypes or defaultProps. Did you accidentally pass a React component?");var r={$$typeof:p,render:e};{var t;Object.defineProperty(r,"displayName",{enumerable:!1,configurable:!0,get:function(){return t},set:function(n){t=n,e.displayName==null&&(e.displayName=n)}})}return r}var Dr=!1;function $e(e){return!!(typeof e=="string"||typeof e=="function"||e===s.Fragment||e===s.Profiler||e===pe||e===s.StrictMode||e===s.Suspense||e===w||e===he||Dr||typeof e=="object"&&e!==null&&(e.$$typeof===g||e.$$typeof===m||e.$$typeof===b||e.$$typeof===C||e.$$typeof===p||e.$$typeof===de||e.$$typeof===k||e[0]===T))}function Ir(e,r){$e(e)||c("memo: The first argument must be a component. Instead received: %s",e===null?"null":typeof e);var t={$$typeof:m,type:e,compare:r===void 0?null:r};{var n;Object.defineProperty(t,"displayName",{enumerable:!1,configurable:!0,get:function(){return n},set:function(a){n=a,e.displayName==null&&(e.displayName=a)}})}return t}function P(){var e=ye.current;if(e===null)throw Error(\`Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://reactjs.org/link/invalid-hook-call for tips about how to debug and fix this problem.\`);return e}function $r(e,r){var t=P();if(r!==void 0&&c("useContext() second argument is reserved for future use in React. Passing it is not supported. You passed: %s.%s",r,typeof r=="number"&&Array.isArray(arguments[2])?\`

Did you call array.map(useContext)? Calling Hooks inside a loop is not supported. Learn more at https://reactjs.org/link/rules-of-hooks\`:""),e._context!==void 0){var n=e._context;n.Consumer===e?c("Calling useContext(Context.Consumer) is not supported, may cause bugs, and will be removed in a future major release. Did you mean to call useContext(Context) instead?"):n.Provider===e&&c("Calling useContext(Context.Provider) is not supported. Did you mean to call useContext(Context) instead?")}return t.useContext(e,r)}function Nr(e){var r=P();return r.useState(e)}function Ur(e,r,t){var n=P();return n.useReducer(e,r,t)}function Wr(e){var r=P();return r.useRef(e)}function Mr(e,r){var t=P();return t.useEffect(e,r)}function Yr(e,r){var t=P();return t.useLayoutEffect(e,r)}function Vr(e,r){var t=P();return t.useCallback(e,r)}function Lr(e,r){var t=P();return t.useMemo(e,r)}function xr(e,r,t){var n=P();return n.useImperativeHandle(e,r,t)}function Fr(e,r){{var t=P();return t.useDebugValue(e,r)}}var M=0,Ne,Ue,We,Me,Ye,Ve,Le;function xe(){}xe.__reactDisabledLog=!0;function qr(){{if(M===0){Ne=console.log,Ue=console.info,We=console.warn,Me=console.error,Ye=console.group,Ve=console.groupCollapsed,Le=console.groupEnd;var e={configurable:!0,enumerable:!0,value:xe,writable:!0};Object.defineProperties(console,{info:e,log:e,warn:e,error:e,group:e,groupCollapsed:e,groupEnd:e})}M++}}function zr(){{if(M--,M===0){var e={configurable:!0,enumerable:!0,writable:!0};Object.defineProperties(console,{log:_({},e,{value:Ne}),info:_({},e,{value:Ue}),warn:_({},e,{value:We}),error:_({},e,{value:Me}),group:_({},e,{value:Ye}),groupCollapsed:_({},e,{value:Ve}),groupEnd:_({},e,{value:Le})})}M<0&&c("disabledDepth fell below zero. This is a bug in React. Please file an issue.")}}var oe=U.ReactCurrentDispatcher,ue;function F(e,r,t){{if(ue===void 0)try{throw Error()}catch(a){var n=a.stack.trim().match(/\\n( *(at )?)/);ue=n&&n[1]||""}return\`
\`+ue+e}}var ie=!1,q;{var Hr=typeof WeakMap=="function"?WeakMap:Map;q=new Hr}function Fe(e,r){if(!e||ie)return"";{var t=q.get(e);if(t!==void 0)return t}var n;ie=!0;var a=Error.prepareStackTrace;Error.prepareStackTrace=void 0;var i;i=oe.current,oe.current=null,qr();try{if(r){var u=function(){throw Error()};if(Object.defineProperty(u.prototype,"props",{set:function(){throw Error()}}),typeof Reflect=="object"&&Reflect.construct){try{Reflect.construct(u,[])}catch(E){n=E}Reflect.construct(e,[],u)}else{try{u.call()}catch(E){n=E}e.call(u.prototype)}}else{try{throw Error()}catch(E){n=E}e()}}catch(E){if(E&&n&&typeof E.stack=="string"){for(var o=E.stack.split(\`
\`),f=n.stack.split(\`
\`),l=o.length-1,v=f.length-1;l>=1&&v>=0&&o[l]!==f[v];)v--;for(;l>=1&&v>=0;l--,v--)if(o[l]!==f[v]){if(l!==1||v!==1)do if(l--,v--,v<0||o[l]!==f[v]){var h=\`
\`+o[l].replace(" at new "," at ");return typeof e=="function"&&q.set(e,h),h}while(l>=1&&v>=0);break}}}finally{ie=!1,oe.current=i,zr(),Error.prepareStackTrace=a}var R=e?e.displayName||e.name:"",y=R?F(R):"";return typeof e=="function"&&q.set(e,y),y}function qe(e,r,t){return Fe(e,!1)}function Kr(e){var r=e.prototype;return!!(r&&r.isReactComponent)}function z(e,r,t){if(e==null)return"";if(typeof e=="function")return Fe(e,Kr(e));if(typeof e=="string")return F(e);switch(e){case s.Suspense:return F("Suspense");case w:return F("SuspenseList")}if(typeof e=="object")switch(e.$$typeof){case p:return qe(e.render);case m:return z(e.type,r,t);case k:return qe(e._render);case g:{var n=e,a=n._payload,i=n._init;try{return z(i(a),r,t)}catch{}}}return""}var ze={},He=U.ReactDebugCurrentFrame;function H(e){if(e){var r=e._owner,t=z(e.type,e._source,r?r.type:null);He.setExtraStackFrame(t)}else He.setExtraStackFrame(null)}function Gr(e,r,t,n,a){{var i=Function.call.bind(Object.prototype.hasOwnProperty);for(var u in e)if(i(e,u)){var o=void 0;try{if(typeof e[u]!="function"){var f=Error((n||"React class")+": "+t+" type \`"+u+"\` is invalid; it must be a function, usually from the \`prop-types\` package, but received \`"+typeof e[u]+"\`.This often happens because of typos such as \`PropTypes.function\` instead of \`PropTypes.func\`.");throw f.name="Invariant Violation",f}o=e[u](r,u,n,t,null,"SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED")}catch(l){o=l}o&&!(o instanceof Error)&&(H(a),c("%s: type specification of %s \`%s\` is invalid; the type checker function must return \`null\` or an \`Error\` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).",n||"React class",t,u,typeof o),H(null)),o instanceof Error&&!(o.message in ze)&&(ze[o.message]=!0,H(a),c("Failed %s type: %s",t,o.message),H(null))}}}function I(e){if(e){var r=e._owner,t=z(e.type,e._source,r?r.type:null);be(t)}else be(null)}var se;se=!1;function Ke(){if(O.current){var e=j(O.current.type);if(e)return\`

Check the render method of \\\`\`+e+"\`."}return""}function Br(e){if(e!==void 0){var r=e.fileName.replace(/^.*[\\\\\\/]/,""),t=e.lineNumber;return\`

Check your code at \`+r+":"+t+"."}return""}function Xr(e){return e!=null?Br(e.__source):""}var Ge={};function Jr(e){var r=Ke();if(!r){var t=typeof e=="string"?e:e.displayName||e.name;t&&(r=\`

Check the top-level render call using <\`+t+">.")}return r}function Be(e,r){if(!(!e._store||e._store.validated||e.key!=null)){e._store.validated=!0;var t=Jr(r);if(!Ge[t]){Ge[t]=!0;var n="";e&&e._owner&&e._owner!==O.current&&(n=" It was passed a child from "+j(e._owner.type)+"."),I(e),c('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.',t,n),I(null)}}}function Xe(e,r){if(typeof e=="object"){if(Array.isArray(e))for(var t=0;t<e.length;t++){var n=e[t];D(n)&&Be(n,r)}else if(D(e))e._store&&(e._store.validated=!0);else if(e){var a=ge(e);if(typeof a=="function"&&a!==e.entries)for(var i=a.call(e),u;!(u=i.next()).done;)D(u.value)&&Be(u.value,r)}}}function Je(e){{var r=e.type;if(r==null||typeof r=="string")return;var t;if(typeof r=="function")t=r.propTypes;else if(typeof r=="object"&&(r.$$typeof===p||r.$$typeof===m))t=r.propTypes;else return;if(t){var n=j(r);Gr(t,e.props,"prop",n,e)}else if(r.PropTypes!==void 0&&!se){se=!0;var a=j(r);c("Component %s declared \`PropTypes\` instead of \`propTypes\`. Did you misspell the property assignment?",a||"Unknown")}typeof r.getDefaultProps=="function"&&!r.getDefaultProps.isReactClassApproved&&c("getDefaultProps is only used on classic React.createClass definitions. Use a static property named \`defaultProps\` instead.")}}function Qr(e){{for(var r=Object.keys(e.props),t=0;t<r.length;t++){var n=r[t];if(n!=="children"&&n!=="key"){I(e),c("Invalid prop \`%s\` supplied to \`React.Fragment\`. React.Fragment can only have \`key\` and \`children\` props.",n),I(null);break}}e.ref!==null&&(I(e),c("Invalid attribute \`ref\` supplied to \`React.Fragment\`."),I(null))}}function Qe(e,r,t){var n=$e(e);if(!n){var a="";(e===void 0||typeof e=="object"&&e!==null&&Object.keys(e).length===0)&&(a+=" You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");var i=Xr(r);i?a+=i:a+=Ke();var u;e===null?u="null":Array.isArray(e)?u="array":e!==void 0&&e.$$typeof===S?(u="<"+(j(e.type)||"Unknown")+" />",a=" Did you accidentally export a JSX literal instead of a component?"):u=typeof e,c("React.createElement: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s",u,a)}var o=mr.apply(this,arguments);if(o==null)return o;if(n)for(var f=2;f<arguments.length;f++)Xe(arguments[f],e);return e===s.Fragment?Qr(o):Je(o),o}var Ze=!1;function Zr(e){var r=Qe.bind(null,e);return r.type=e,Ze||(Ze=!0,W("React.createFactory() is deprecated and will be removed in a future major release. Consider using JSX or use React.createElement() directly instead.")),Object.defineProperty(r,"type",{enumerable:!1,get:function(){return W("Factory.type is deprecated. Access the class directly before passing it to createFactory."),Object.defineProperty(this,"type",{value:e}),e}}),r}function et(e,r,t){for(var n=yr.apply(this,arguments),a=2;a<arguments.length;a++)Xe(arguments[a],n.type);return Je(n),n}try{var er=Object.freeze({})}catch{}var rt=Qe,tt=et,nt=Zr,at={map:x,forEach:Cr,count:Er,toArray:wr,only:Pr};s.Children=at,s.Component=A,s.PureComponent=Z,s.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED=U,s.cloneElement=tt,s.createContext=Or,s.createElement=rt,s.createFactory=nt,s.createRef=lr,s.forwardRef=Ar,s.isValidElement=D,s.lazy=Sr,s.memo=Ir,s.useCallback=Vr,s.useContext=$r,s.useDebugValue=Fr,s.useEffect=Mr,s.useImperativeHandle=xr,s.useLayoutEffect=Yr,s.useMemo=Lr,s.useReducer=Ur,s.useRef=Wr,s.useState=Nr,s.version=G})()}(le)),le}var ur;function ft(){return ur||(ur=1,fe.exports=ct()),fe.exports}ft();const lt="test",vt="test";console.log({rackEnv:lt,railsEnv:vt});
"
`;

exports[`erb > typescript and sass 1`] = `
"const e="development",n="development";console.log({rackEnv:e,railsEnv:n});
"
`;
